test/validation/api/pktio/pktio.c

```diff
@@ -2578,6 +2578,152 @@ static void pktio_test_extra_stats(void)
+}
+
+static void pktio_test_proto_statistics_counters(void)
+{
+	odp_pktio_t pktio_rx, pktio_tx;
+	odp_pktio_t pktio[MAX_NUM_IFACES] = {
+		ODP_PKTIO_INVALID, ODP_PKTIO_INVALID
+	};
+	odp_packet_t pkt;
+	odp_packet_t tx_pkt[1000];
+	uint32_t pkt_seq[1000];
+	odp_event_t ev;
+	int i, pkts, tx_pkts, ret, alloc = 0;
+	odp_pktout_queue_t pktout;
+	uint64_t wait = odp_schedule_wait_time(ODP_TIME_MSEC_IN_NS);
+	odp_proto_stats_capability_t capa;
+	odp_packet_proto_stats_opt_t opt;
+	odp_proto_stats_param_t param;
+	odp_proto_stats_data_t data;
+	odp_pktio_config_t config;
+	odp_proto_stats_t stat;
+
+	memset(&pktout, 0, sizeof(pktout));
+	odp_pktio_config_init(&config);
+	config.pktout.bit.proto_stats_ena = 1;
+
+	for (i = 0; i < num_ifaces; i++) {
+		pktio[i] = create_pktio(i, ODP_PKTIN_MODE_SCHED,
+					ODP_PKTOUT_MODE_DIRECT);
+
+		CU_ASSERT_FATAL(pktio[i] != ODP_PKTIO_INVALID);
+		odp_pktio_config(pktio[i], &config);
+	}
+	pktio_tx = pktio[0];
+	pktio_rx = (num_ifaces > 1) ? pktio[1] : pktio_tx;
+
+	CU_ASSERT_FATAL(odp_pktout_queue(pktio_tx, &pktout, 1) == 1);
+
+	ret = odp_pktio_start(pktio_tx);
+	CU_ASSERT(ret == 0);
+	if (num_ifaces > 1) {
+		ret = odp_pktio_start(pktio_rx);
+		CU_ASSERT(ret == 0);
+	}
+
+	odp_proto_stats_param_init(&param);
+	odp_proto_stats_capability(pktio_tx, &capa);
+	CU_ASSERT(capa.tx.counters.all_bits != 0);
+	param.counters.all_bits = capa.tx.counters.all_bits;
+	/* Create statistics object with all supported counters */
+	stat = odp_proto_stats_create("flow0_stat", &param);
+	CU_ASSERT_FATAL(stat != ODP_PROTO_STATS_INVALID);
+
+	alloc = create_packets(tx_pkt, pkt_seq, 1000, pktio_tx, pktio_rx);
+
+	/* Attach statistics object to all Tx packets */
+	for (pkts = 0; pkts < alloc; pkts++) {
+		opt.stat = stat;
+		opt.oct_count0_adj = 0;
+		/* oct1 contains byte count of packets excluding Ethernet header */
+		opt.oct_count1_adj = -14;
+		odp_packet_proto_stats_request(tx_pkt[pkts], &opt);
+	}
+
+	/* send */
+	for (pkts = 0; pkts != alloc; ) {
+		ret = odp_pktout_send(pktout, &tx_pkt[pkts], alloc - pkts);
+		if (ret < 0) {
+			CU_FAIL("unable to send packet\n");
+			break;
+		}
+		pkts += ret;
+	}
+	tx_pkts = pkts;
+
+	/* get */
+	for (i = 0, pkts = 0; i < 1000 && pkts != tx_pkts; i++) {
+		ev = odp_schedule(NULL, wait);
+		if (ev != ODP_EVENT_INVALID) {
+			if (odp_event_type(ev) == ODP_EVENT_PACKET) {
+				pkt = odp_packet_from_event(ev);
+				if (pktio_pkt_seq(pkt) != TEST_SEQ_INVALID)
+					pkts++;
+			}
+			odp_event_free(ev);
+		}
+	}
+
+	CU_ASSERT(pkts == tx_pkts);
+
+	ret = odp_proto_stats(stat, &data);
+	CU_ASSERT(ret == 0);
+	CU_ASSERT(!(capa.tx.counters.bit.tx_pkts && (data.tx_pkts != (uint64_t)pkts)));
+	CU_ASSERT(!(capa.tx.counters.bit.tx_pkt_drops && (data.tx_pkt_drops > 0)));
+	CU_ASSERT(!(capa.tx.counters.bit.tx_oct_count0_drops && (data.tx_oct_count0_drops > 0)));
+	CU_ASSERT(!(capa.tx.counters.bit.tx_oct_count1_drops && (data.tx_oct_count1_drops > 0)));
+	if (capa.tx.oct_count0_adj) {
```

Comment:
I think these tx octest counter tests are not very readable. Something like this would seem better, wouldn't it:
```
        if (capa.tx.counters.bit.tx_oct_count0) {
                uint64_t counted_bytes = PKT_LEN_NORMAL;

                if (capa.tx.oct_count0_adj)
                        counted_bytes += opt.oct_count0_adj;
                CU_ASSERT(data.tx_oct_count0 != counted_bytes * (uint64_t)pkts);
        }
```

Reply:
Thanks for suggestion, will address in next version.